geopandas/_block.py

```diff
@@ -0,0 +1,347 @@
+
+import numpy as np
+
+from pandas.core.internals import Block, NonConsolidatableMixIn
+from pandas.core.common import is_null_slice
+
+
+class GeometryBlock(NonConsolidatableMixIn, Block):
+    """ implement a datetime64 block with a tz attribute """
+    __slots__ = ()
+
+    @property
+    def _holder(self):
+        from geopandas.vectorized import VectorizedGeometry
+        return VectorizedGeometry
+
+    def __init__(self, values, placement, ndim=2, **kwargs):
+
+        if not isinstance(values, self._holder):
+            raise TypeError("values must be a VectorizedGeometry object")
+
+        super(GeometryBlock, self).__init__(values, placement=placement,
+                                            ndim=ndim, **kwargs)
+
+    @property
+    def _box_func(self):
+        # TODO does not seems to be used at the moment (from the examples) ?
+        print("I am boxed")
+        from shapely.geometry.base import geom_factory
+        return geom_factory
+
+    # @property
+    # def _na_value(self):
+    #     return None
+    #
+    # @property
+    # def fill_value(self):
+    #     return tslib.iNaT
+
+    # TODO
+    # def copy(self, deep=True, mgr=None):
+    #     """ copy constructor """
+    #     values = self.values
+    #     if deep:
+    #         values = values.copy(deep=True)
+    #     return self.make_block_same_class(values)
+
+    def external_values(self):
+        """ we internally represent the data as a DatetimeIndex, but for
+        external compat with ndarray, export as a ndarray of Timestamps
+        """
+        #return np.asarray(self.values)
+        print("I am densified (external_values, {} elements)".format(len(self)))
+        return self.values.to_dense()
+
+    def formatted_values(self, dtype=None):
+        """ return an internal format, currently just the ndarray
+        this should be the pure internal API format
+        """
+        return self.to_dense()
+
+    def to_dense(self):
+        print("I am densified ({} elements)".format(len(self)))
+        return self.values.to_dense().view()
+
+    def __getitem__(self, key):
+        values = self.values[key]
+        return GeometryBlock(values, placement=slice(0, len(values), 1),
```

Comment:
@mrocklin for which test /functionality was this needed (the densification during slicing / alignment ?) because it breaks some other things